storage.bs

```diff
@@ -94,37 +100,292 @@ function retrieveNextChunk(nextChunkInfo) {
+stored lives. User agents are expected to store this data, and make it available across <a>agent</a>
+and even <a>agent cluster</a> boundaries, in an <a>implementation-defined</a> manner, so that this
+standard and standards using this standard can access the contents.
+
+<hr>
+
+<p>To <dfn>obtain a storage bottle map</dfn>, given a <a>storage type</a> <var>type</var>,
+<a for=/>environment settings object</a> <var>environment</var>, and <a>storage identifier</a>
+<var>identifier</var>, run these steps:</p>
+
+<ol>
+ <li><p>Let <var>shed</var> be null.
+
+ <li><p>If <var>type</var> is "<code>local</code>", then set <var>shed</var> to the user agent's
+ <a for="user agent">storage shed</a>.
+
+ <li>
+  <p>Otherwise:
+
+  <ol>
+   <li><p>Assert: <var>type</var> is "<code>session</code>".
+
+   <li><p>Set <var>shed</var> to <var>environment</var>'s
+   <a class=XXX href=#lack-of-browsing-session>browsing session</a>'s
+   <a for="browsing session">storage shed</a>.
+  </ol>
+
+ <li><p>Let <var>shelf</var> be the result of running <a>obtain a storage shelf</a>, with
+ <var>shed</var>, <var>environment</var>, and <var>type</var>.
+
+ <li><p>If <var>shelf</var> is failure, then return failure.
+
+ <li><p>Let <var>bucket</var> be <var>shelf</var>'s <a>bucket map</a>["<code>default</code>"].
+
+ <li><p>Let <var>bottle</var> be <var>bucket</var>'s <a>bottle map</a>[<var>identifier</var>].
+
+ <li><p>Let <var>proxyMap</var> be a new <a>storage proxy map</a> whose
+ <a for="storage proxy map">backing map</a> is <var>bottle</var>'s <a for="storage bottle">map</a>.
+
+ <li><p><a for=set>Append</a> <var>proxyMap</var> to <var>bottle</var>'s
+ <a for="storage bottle">proxy map reference set</a>.
+
+ <li><p>Return <var>proxyMap</var>.
+</ol>
+
+<p>To <dfn export>obtain a local storage bottle map</dfn>, given an
+<a for=/>environment settings object</a> <var>environment</var> and <a>storage identifier</a>
+<var>identifier</var>, return the result of running <a>obtain a storage bottle map</a> with
+"<code>local</code>", <var>environment</var>, and <var>identifier</var>.
+
+<p>To <dfn export>obtain a session storage bottle map</dfn>, given an
+<a for=/>environment settings object</a> <var>environment</var> and <a>storage identifier</a>
+<var>identifier</var>, return the result of running <a>obtain a storage bottle map</a> with
+"<code>session</code>", <var>environment</var>, and <var>identifier</var>.
+
+
+<h3 id=storage-proxy-maps>Storage proxy maps</h3>
+
+<p>A <dfn>storage proxy map</dfn> is equivalent to a <a for=/>map</a>, except that all operations
+are instead performed on its <dfn for="storage proxy map">backing map</dfn>.
+
+<p class="XXX">This allows for the <a for="storage proxy map">backing map</a> to be replaced. This
+is needed for <a href="https://github.com/whatwg/storage/issues/4">issue #4</a> and potentially the
+<a href="https://privacycg.github.io/storage-access/">Storage Access API</a>.
```

Comment:
I guess maybe it'll be used as part of the future infrastructure, but right now the "proxy map reference set" is append-only, and never read from.

Reply:
I also share this concern, although it seems like that's a future-work follow-up on the "replace" algorithm proposed at the bottom of https://github.com/whatwg/storage/issues/18#issuecomment-614751336 ?

I'm assuming that's what's going to happen, in which case we might as well wait, as it seems likely to be non-trivial.

Reply:
Yeah you need to know who have a proxy map in order to let their proxies point elsewhere when storage is replaced.

Reply:
I'm a bit surprised by this feedback though as I thought from #18 that we agreed on this approach.